citro3d/src/lib.rs

```diff
@@ -130,218 +114,30 @@ impl Instance {
         render::Target::new(width, height, screen, depth_format, Rc::clone(&self.queue))
     }
 
-    /// Select the given render target for drawing the frame. This must be called
-    /// as pare of a render call (i.e. within the call to
-    /// [`render_frame_with`](Self::render_frame_with)).
-    ///
-    /// # Errors
-    ///
-    /// Fails if the given target cannot be used for drawing, or called outside
-    /// the context of a frame render.
-    #[doc(alias = "C3D_FrameDrawOn")]
-    pub fn select_render_target(&mut self, target: &render::Target<'_>) -> Result<()> {
-        let _ = self;
-        if unsafe { citro3d_sys::C3D_FrameDrawOn(target.as_raw()) } {
-            Ok(())
-        } else {
-            Err(Error::InvalidRenderTarget)
-        }
-    }
-
     /// Render a frame. The passed in function/closure can mutate the instance,
     /// such as to [select a render target](Self::select_render_target)
     /// or [bind a new shader program](Self::bind_program).
     #[doc(alias = "C3D_FrameBegin")]
     #[doc(alias = "C3D_FrameEnd")]
-    pub fn render_frame_with(&mut self, f: impl FnOnce(&mut Self)) {
+    pub fn render_frame_with<'istance: 'frame, 'frame>(
+        &'istance mut self,
+        f: impl FnOnce(RenderPass<'frame>) -> RenderPass<'frame>,
+    ) {
         unsafe {
             citro3d_sys::C3D_FrameBegin(
```

Comment:
imo it would be better to have a RAII object called something like `C3DFrame` or just `Frame` that does `FrameBegin` in `new` and called `FrameEnd` in `Drop`, then have `RenderPass` hold an instance of this object, so it cannot be constructed without having the frame managed correctly (given render pass has the invariant of having an active frame for using it to be valid)

Reply:
I like the idea, but `Drop` order would be an issue if the `RenderPass` is meant to be dropped within the frame function.